src/js/trove/load-lib.js

```diff
@@ -310,31 +345,48 @@
         }
       };
 
-
-      return runtime.pauseStack(function(restarter) {
+      var pauseFunction = function(restarter) {
         var mainReached = false;
         var mainResult = "Main result unset: should not happen";
         postLoadHooks[main] = function(answer) {
           mainReached = true;
           mainResult = answer;
-        }
-        return otherRuntime.runThunk(function() {
+        };
+
+        var thunk = function() {
           otherRuntime.modules = realm;
           return otherRuntime.runStandalone(staticModules, realm, depMap, toLoad, postLoadHooks);
-        }, function(result) {
+        };
+
+        var then = function(result) {
+          var retVal;
+          
           if(!mainReached) {
             // NOTE(joe): we should only reach here if there was an error earlier
             // on in the chain of loading that stopped main from running
-            restarter.resume(makeModuleResult(otherRuntime, result, makeRealm(realm), runtime.nothing));
+            retVal = makeModuleResult(otherRuntime, result, makeRealm(realm), runtime.nothing);
           }
           else {
             var finalResult = otherRuntime.makeSuccessResult(mainResult);
             finalResult.stats = result.stats;
-            restarter.resume(makeModuleResult(otherRuntime, finalResult, makeRealm(realm), runtime.nothing));
+            retVal = makeModuleResult(otherRuntime, finalResult, makeRealm(realm), runtime.nothing);
           }
-        });
-      });
 
+          if (runtime.bounceAllowed) {
+            restarter.resume(retVal);
+          } else {
+            return retVal;
+          }
```

Comment:
I added a dummy restarter and did what you described (I think) and the code in load-lib is quite a bit cleaner now.